metricflow/inference/rule/base.py

```diff
@@ -0,0 +1,139 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from enum import Enum
+import logging
+import inspect
+from typing import List, Optional, Tuple, Type
+
+from metricflow.dataflow.sql_column import SqlColumn
+from metricflow.inference.context.base import InferenceContext
+
+
+logger = logging.getLogger(__name__)
+
+
+class InferenceSignalConfidence(Enum):
+    """A discrete enumeration of possible confidence values for an inference signal.
+
+    We chose discrete confidence values instead of a continuous range (e.g a float between 0 and 1)
+    to standardize confidence outputs between heuristic rules. We want to avoid different rules
+    assuming different values for, say, "medium" or "high" confidence, since this could skew
+    results in favor/against certain rules.
+    """
+
+    FOR_SURE = 3
+    HIGH = 2
+    MEDIUM = 1
+    LOW = 0
+
+
+class InferenceSignalType(str, Enum):
+    """The type of the inferred signal about a column. Effectively, this represents what a rule can think a column is."""
+
+    # identifiers
+    IDENTIFER = "identifier"
+    PRIMARY_IDENTIFIER = "identifier_primary"
+    FOREIGN_IDENTIFIER = "identifier_foreign"
+
+    # dimension detection
+    DIMENSION = "dimension"
+    TIME_DIMENSION = "dimension_time"
+    CATEGORICAL_DIMENSION = "dimension_categorical"
+
+    # measure fields, i.e, not an ID nor a dimension
+    MEASURE_FIELD = "measure_field"
+
+    @staticmethod
+    def conflict(a: InferenceSignalType, b: InferenceSignalType) -> bool:
+        """Helper function for determining whether two signal types conflict with each other.
+
+        This allows us to implement broad signal types with specializations. It makes it possible
+        for rules to produce results such as "I think this is an identifier, but I don't know
+        whether it is primary or foreign".
+
+        Example: A column cannot be both PRIMARY_IDENTIFIER and a FOREIGN_IDENTIFIER,
+        so there is a conflict in this case. However, a column that is a TIME_DIMENSION is still
+        a DIMENSION, so no conflict is found in that case.
+
+        NOTE: The implementation assumes that broader types are prefixes of more specific types,
+        e.g, "identifier" (generic) is a prefix of "identifier_primary" (specific).
+        """
+        return not a.value.startswith(b.value) and not b.value.startswith(a.value)
+
+
+@dataclass(frozen=True)
+class InferenceSignal:
+    """Encapsulates a piece of evidence about a column produced by an inference rule.
+
+    column: the target column for this signal
+    type: the type of the signal
+    reason: a human-readable string that explains why this signal was produced. It may
+        eventually reach the user's eyeballs.
+    confidence: the confidence that this signal is correct.
+    """
+
+    column: SqlColumn
+    type: InferenceSignalType
```

Comment:
In thinking about this a bit further I think you want a hierarchy around InferenceSignalType that makes sense as a return.

Fundamentally, you're trying to figure out how to populate field sets in a DataSource. That means you care about, basically, three things: Dimensions, Identifiers, and Measures

So you can have essentially three concrete classes that are something like `InferredColumnType` and each of those can then encapsulate their own hierarchy of inferred attributes, and manage consistency checking internally. They'd implement a shared `InferredColumnType` interface, so you can operate on them interchangeably.

If you ever have something that's shared across those class divisions - like maybe there's a signal that something maybe shouldn't be included in the dimension/identifier/measure set but if the user does want it the inference should remain - you can define another cross-cutting class that essentially extends the functionality across all `InferredColumnType` base classes.

That gives you a much cleaner, more extensible model, full typechecking support, and will naturally dictate the boundaries of what's acceptable in terms of return types for various inference rules. The internals of which class of thing to return might be a little more difficult on the implementer, but everything else will likely be cleaner, and you can smooth out those rough edges as needed.